Code/GraphMol/Depictor/RDDepictor.cpp

```diff
@@ -427,4 +428,163 @@ unsigned int compute2DCoordsMimicDistMat(
   unsigned int cid = copyCoordinate(mol, efrags, clearConfs);
   return cid;
 }
+
+//! \brief Compute 2D coordinates where a piece of the molecule is
+//   constrained to have the same coordinates as a reference.
+/*!
+  This function generates a depiction for a molecule where a piece of the
+  molecule is constrained to have the same coordinates as a reference.
+
+  This is useful for, for example, generating depictions of SAR data
+  sets so that the cores of the molecules are all oriented the same way.
+
+  ARGUMENTS:
+
+  \param mol -    the molecule to be aligned, this will come back
+                  with a single conformer.
+  \param reference -    a molecule with the reference atoms to align to;
+                        this should have a depiction.
+  \param confId -       (optional) the id of the reference conformation to use
+  \param referencePattern -  (optional) an optional molecule to be used to
+                             generate the atom mapping between the molecule
+                             and the reference.
+  \param acceptFailure - (optional) if True, standard depictions will be generated
+                         for molecules that don't have a substructure match to the
+                         reference; if false, throws a DepictException.
+
+*/
+  
+void generateDepictionMatching2DStructure( RDKit::ROMol &mol ,
+					   RDKit::ROMol &reference ,
+					   int confId ,
+					   RDKit::ROMol *referencePattern ,
+					   bool acceptFailure ) {
+
+  std::vector<int> refMatch;
+  RDKit::MatchVectType matchVect;
+  if( referencePattern ) {
+    if( reference.getNumAtoms( true ) != referencePattern->getNumAtoms( true ) ) {
+      throw RDDepict::DepictException( "When a pattern is provided, it must have the same number of atoms as the reference" );
+    }
+    RDKit::MatchVectType refMatchVect;
+    RDKit::SubstructMatch( reference , *referencePattern , refMatchVect );
+    if( refMatchVect.empty() ) {
+      throw RDDepict::DepictException( "Reference does not map to itself." );
+    }
+    refMatch.reserve( refMatchVect.size() );
+    for( size_t i = 0 , is = refMatchVect.size() ; i < is ; ++i ) {
+      refMatch.push_back( refMatchVect[i].second );
+    }
+    RDKit::SubstructMatch( mol , *referencePattern , matchVect );
+  } else {
+    refMatch.reserve( reference.getNumAtoms( true ) );
+    for( int i = 0 , is = reference.getNumAtoms( true ) ; i < is ; ++i ) {
```

Comment:
I grant you that this is a slightly odd style, but I do like it.  getNumAtoms() is only called once, at loop initialisation, which should have a slight efficiency gain.  Pushing is outside the loop pollutes the wider scope.  Initialising it in the loop header at the same time as i makes it very clear that it is only being used inside the loop.  I will use the second option, and maybe the compiler will optimise it out.